File: backend-api/app/schemas/soc2.py
"""Response contract for the SOC 2 report projection.

Every model here is a *projection* of two frozen inputs: the mapping pinned on
the scan and the results that scan actually produced. Nothing in this contract
lets a rating be computed, promoted or inferred - ``configuration_rating`` is
copied verbatim from the pinned mapping and is accompanied by
``rating_is_computed: false`` so a consumer cannot mistake it for a score.

Coverage is always reported as a numerator and a denominator with the percentage
alongside, never as a bare percentage, because "100%" of a two-control sample is
not the same claim as "100%" of a forty-control sample.
"""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

# A rating is whatever the pinned mapping's own ``rating_vocabulary`` allows.
# It is intentionally untyped here: this contract transcribes GRC judgments, it
# does not define them, and a future approved mapping may use other words.
Soc2Rating = str

ProjectionStatus = Literal["available", "no_projection"]

AssessmentCompleteness = Literal[
    "complete",  # every applicable control in this point of focus was assessed
    "partial",  # some applicable controls produced no assessed result
    "not_assessed",  # applicable controls exist, none were assessed
    "not_applicable",  # this scan selected none of the mapped controls
    "no_automated_coverage",  # the mapping maps this point to no automation
]

NOT_A_CERTIFICATION = (
    "This is not a SOC 2 certification, examination, attestation or audit "
    "opinion, and it is not evidence that any criterion is met. It reports "
    "Microsoft 365 configuration results projected through a crosswalk that "
    "has not been approved by a GRC reviewer."
)

RATING_OWNERSHIP = (
    "Configuration ratings are human-owned GRC judgments copied verbatim from "
    "the pinned mapping. AutoAudit never creates, computes, promotes or lowers "
    "a rating, and scan results never change one."
)


class Soc2Approval(BaseModel):
    """The pinned mapping's approval block, reproduced verbatim."""

    model_config = ConfigDict(extra="allow")

    approved: bool = False
    reviewer_name: str | None = None
    reviewer_role: str | None = None
    approved_at: str | None = None
    decision_reference: str | None = None
    note: str | None = None


class Soc2Identity(BaseModel):
    """The SOC 2 side of the mapping, reproduced verbatim."""

    model_config = ConfigDict(extra="allow")

    framework: str | None = None
    criteria_family: str | None = None
    authoritative_criteria: list[str] = Field(default_factory=list)


class Soc2ReportHeader(BaseModel):
    """Document-level header. Carries the approval state and the disclaimer."""

    mapping_id: str
    mapping_version: str
    mapping_digest: str
    mapping_status: str | None = None
    schema_version: int | None = None
    soc2: Soc2Identity
    benchmark: dict[str, Any] = Field(default_factory=dict)
    source: dict[str, Any] | None = None
    rating_vocabulary: list[str] = Field(default_factory=list)
    approval: Soc2Approval
    approval_pending: bool
    not_a_certification: str = NOT_A_CERTIFICATION
    rating_ownership: str = RATING_OWNERSHIP


class Soc2Provenance(BaseModel):
    """Everything needed to reproduce this document from frozen inputs."""

    scan_id: int
    scan_status: str
    started_at: datetime | None = None
    finished_at: datetime | None = None
    framework: str
    benchmark: str
    benchmark_version: str
    metadata_digest: str | None = None
    policy_corpus_digest: str | None = None
    mapping_id: str | None = None
    mapping_version: str | None = None
    mapping_digest: str | None = None
    correlation_id: str | None = None
    semantics_version: str | None = None
    lifecycle_version: str | None = None
    evidence_version: str | None = None
    connection_snapshot_present: bool = False
    rendered_from: Literal["pinned_mapping_snapshot"] = "pinned_mapping_snapshot"
    generated_at: datetime


class Soc2ControlEvidence(BaseModel):
    """One mapped CIS control as this scan actually found it."""

    control_id: str
    title: str | None = None
    # Whether the scan's pinned benchmark metadata defines this control at all.
    in_benchmark: bool = False
    # Whether the scan holds a result row for it.
    in_scan: bool
    selected: bool | None = None
    # A scan result status, or one of two synthetic ones. ``missing_result`` means
    # the pinned benchmark defines the control but the scan has no row - an
    # integrity gap that stays in the coverage denominator. ``not_in_benchmark``
    # means the mapping cites a control this benchmark never had. Neither is ever
    # a pass, and neither is ever a tenant failure.
    status: str
    reason_code: str | None = None
    automation_status: str | None = None
    # Execution provenance only. Collected tenant payloads are never included.
    provenance: dict[str, Any] | None = None


class Soc2Coverage(BaseModel):
    """Coverage with both terms exposed. Never a bare percentage."""

    mapped_control_count: int
    in_scan_count: int
    # Denominator: mapped controls this scan actually selected.
    applicable_count: int
    # Numerator: of those, the ones that produced passed or failed.
    assessed_count: int
    passed_count: int
    failed_count: int
    indeterminate_count: int
    error_count: int
    not_assessable_count: int
    pending_count: int
    # Controls the pinned benchmark defines but the scan holds no row for. They
    # stay in the denominator: a lost row is unknown, never covered.
    missing_result_count: int = 0
    # Selected-but-unassessed, for readability. Equals applicable - assessed.
    unassessed_count: int
    # Mapped controls the scan did not select, plus mapped controls the pinned
    # benchmark does not contain. Out of scope for this scan, not a gap.
    not_in_scope_count: int
    missing_from_benchmark_count: int
    coverage_percent: float | None = None
    compliance_percent: float | None = None
    assessment_completeness: AssessmentCompleteness
    partial_assessment: bool
    coverage_statement: str


class Soc2ManualEvidence(BaseModel):
    """An approved manual or inherited evidence record, reported separately."""

    record_id: int
    control_id: str
    evidence_source: str
    status: str
    control_owner: str | None = None
    evidence_owner: str | None = None
    collection_period_start: datetime | None = None
    collection_period_end: datetime | None = None
    expires_at: datetime | None = None
    cadence: str | None = None
    reviewed_at: datetime | None = None
    current_revision_number: int = 0
    # Structural, not advisory: manual and inherited assurance is a separate
    # stream and can never raise automated configuration coverage.
    counted_in_automated_coverage: Literal[False] = False


class Soc2Limitations(BaseModel):
    """Residual limitation text, verbatim from the pinned mapping."""

    residual_limitation: str | None = None
    residual_scope: str | None = None


class Soc2PointOfFocus(BaseModel):
    """One point of focus: its owned rating beside what the scan found."""

    point_id: str
    criterion: str
    point_of_focus: str
    configuration_rating: Soc2Rating
    rating_source: Literal["pinned_mapping"] = "pinned_mapping"
    rating_is_computed: Literal[False] = False
    evidence_selector: str | None = None
    selector_resolved: bool = True
    mapped_control_ids: list[str] = Field(default_factory=list)
    automated_evidence: list[Soc2ControlEvidence] = Field(default_factory=list)
    coverage: Soc2Coverage
    manual_residual_evidence: list[Soc2ManualEvidence] = Field(default_factory=list)
    limitations: Soc2Limitations


class Soc2CriterionCoverage(BaseModel):
    """Criterion-level classification, verbatim from the pinned mapping."""

    model_config = ConfigDict(extra="allow")

    criterion: str
    configuration_coverage_classification: str | None = None


class Soc2Totals(BaseModel):
    """Document totals, so a reader can confirm nothing was dropped."""

    points_of_focus_count: int
    unique_mapped_control_ids: int
    rating_counts: dict[str, int] = Field(default_factory=dict)
    criteria_count: int


class Soc2ManualEvidenceStream(BaseModel):
    """Status of the separate manual/inherited evidence stream."""

    available: bool
    approved_record_count: int = 0
    counted_in_automated_coverage: Literal[False] = False
    note: str


class Soc2ReportResponse(BaseModel):
    """The SOC 2 report projection for one scan."""

    scan_id: int
    projection_available: bool
    projection_status: ProjectionStatus
    message: str
    header: Soc2ReportHeader | None = None
    provenance: Soc2Provenance | None = None
    criteria_coverage_summary: list[Soc2CriterionCoverage] = Field(default_factory=list)
    points_of_focus: list[Soc2PointOfFocus] = Field(default_factory=list)
    totals: Soc2Totals | None = None
    manual_evidence_stream: Soc2ManualEvidenceStream | None = None
    # Structural findings against the pinned mapping and pinned metadata.
    # Reported, never acted on: no finding changes a rating or a result.
    mapping_resolution_findings: list[dict[str, str]] = Field(default_factory=list)

File: frontend/src/types/soc2.ts
/**
 * Response contract for `GET /v1/scans/{id}/soc2-report`.
 *
 * These types mirror `backend-api/app/schemas/soc2.py`. Three things they encode
 * deliberately, because the UI must not be able to undo them:
 *
 * - `configuration_rating` is a human-owned GRC judgment copied verbatim from the
 *   pinned crosswalk. `rating_is_computed` is always `false`. Nothing in the scan
 *   result may be used to derive, promote or override it.
 * - coverage always carries `assessed_count` **and** `applicable_count`. A bare
 *   percentage is never a complete statement, so never render one alone.
 * - `manual_residual_evidence` is a separate stream with
 *   `counted_in_automated_coverage: false`. It must not be added to any automated
 *   coverage figure.
 */

/** Whatever the pinned mapping's own `rating_vocabulary` allows (today Yes/Partial/No). */
export type Soc2Rating = string;

export type Soc2ProjectionStatus = 'available' | 'no_projection';

export type Soc2AssessmentCompleteness =
  | 'complete'
  | 'partial'
  | 'not_assessed'
  | 'not_applicable'
  | 'no_automated_coverage';

/** The pinned mapping's approval block, verbatim. `approved` is false today. */
export type Soc2Approval = {
  approved: boolean;
  reviewer_name?: string | null;
  reviewer_role?: string | null;
  approved_at?: string | null;
  decision_reference?: string | null;
  note?: string | null;
};

export type Soc2Identity = {
  framework?: string | null;
  criteria_family?: string | null;
  authoritative_criteria?: string[];
};

export type Soc2ReportHeader = {
  mapping_id: string;
  mapping_version: string;
  mapping_digest: string;
  mapping_status?: string | null;
  schema_version?: number | null;
  soc2: Soc2Identity;
  benchmark?: Record<string, unknown>;
  source?: Record<string, unknown> | null;
  rating_vocabulary?: string[];
  approval: Soc2Approval;
  approval_pending: boolean;
  /** Must be displayed with the report; never suppress it. */
  not_a_certification: string;
  rating_ownership: string;
};

export type Soc2Provenance = {
  scan_id: number;
  scan_status: string;
  started_at?: string | null;
  finished_at?: string | null;
  framework: string;
  benchmark: string;
  benchmark_version: string;
  metadata_digest?: string | null;
  policy_corpus_digest?: string | null;
  mapping_id?: string | null;
  mapping_version?: string | null;
  mapping_digest?: string | null;
  correlation_id?: string | null;
  semantics_version?: string | null;
  lifecycle_version?: string | null;
  evidence_version?: string | null;
  connection_snapshot_present: boolean;
  rendered_from: 'pinned_mapping_snapshot';
  generated_at: string;
};

/** One mapped CIS control as the scan actually found it. */
export type Soc2ControlEvidence = {
  control_id: string;
  title?: string | null;
  /** Whether the scan's pinned benchmark defines this control at all. */
  in_benchmark: boolean;
  /** Whether the scan holds a result row for it. */
  in_scan: boolean;
  selected?: boolean | null;
  /**
   * A scan result status, or one of two synthetic ones. `missing_result` means
   * the pinned benchmark defines the control but the scan has no row - an
   * integrity gap that stays in the coverage denominator. `not_in_benchmark`
   * means the mapping cites a control this benchmark never had. Neither is ever
   * a pass, and neither is ever a tenant failure.
   */
  status: string;
  reason_code?: string | null;
  automation_status?: string | null;
  /** Execution provenance only; collected tenant payloads are never included. */
  provenance?: Record<string, unknown> | null;
};

/** Numerator and denominator are both required. Never render the percent alone. */
export type Soc2Coverage = {
  mapped_control_count: number;
  in_scan_count: number;
  /** Denominator: mapped controls this scan selected. */
  applicable_count: number;
  /** Numerator: of those, the ones that produced passed or failed. */
  assessed_count: number;
  passed_count: number;
  failed_count: number;
  indeterminate_count: number;
  error_count: number;
  not_assessable_count: number;
  pending_count: number;
  /** Controls with no result row at all. Unknown, never covered. */
  missing_result_count: number;
  unassessed_count: number;
  not_in_scope_count: number;
  missing_from_benchmark_count: number;
  coverage_percent?: number | null;
  compliance_percent?: number | null;
  assessment_completeness: Soc2AssessmentCompleteness;
  /** True when some selected control produced no assessment. Label it. */
  partial_assessment: boolean;
  coverage_statement: string;
};

export type Soc2ManualEvidence = {
  record_id: number;
  control_id: string;
  evidence_source: string;
  status: string;
  control_owner?: string | null;
  evidence_owner?: string | null;
  collection_period_start?: string | null;
  collection_period_end?: string | null;
  expires_at?: string | null;
  cadence?: string | null;
  reviewed_at?: string | null;
  current_revision_number: number;
  counted_in_automated_coverage: false;
};

export type Soc2Limitations = {
  residual_limitation?: string | null;
  residual_scope?: string | null;
};

export type Soc2PointOfFocus = {
  point_id: string;
  criterion: string;
  point_of_focus: string;
  /** Copied from the pinned mapping. Scan results never change this. */
  configuration_rating: Soc2Rating;
  rating_source: 'pinned_mapping';
  rating_is_computed: false;
  evidence_selector?: string | null;
  /** False when a selector was not understood; coverage is then incomplete. */
  selector_resolved: boolean;
  mapped_control_ids: string[];
  automated_evidence: Soc2ControlEvidence[];
  coverage: Soc2Coverage;
  manual_residual_evidence: Soc2ManualEvidence[];
  limitations: Soc2Limitations;
};

export type Soc2CriterionCoverage = {
  criterion: string;
  configuration_coverage_classification?: string | null;
};

export type Soc2Totals = {
  points_of_focus_count: number;
  unique_mapped_control_ids: number;
  rating_counts: Record<string, number>;
  criteria_count: number;
};

export type Soc2ManualEvidenceStream = {
  /** False means "could not be read", which is unknown, not none. */
  available: boolean;
  approved_record_count: number;
  counted_in_automated_coverage: false;
  note: string;
};

export type Soc2ResolutionFinding = {
  code: string;
  control_id: string;
  detail: string;
};

export type Soc2ReportResponse = {
  scan_id: number;
  projection_available: boolean;
  projection_status: Soc2ProjectionStatus;
  /** Shown as-is when `projection_available` is false. */
  message: string;
  header?: Soc2ReportHeader | null;
  provenance?: Soc2Provenance | null;
  criteria_coverage_summary?: Soc2CriterionCoverage[];
  points_of_focus?: Soc2PointOfFocus[];
  totals?: Soc2Totals | null;
  manual_evidence_stream?: Soc2ManualEvidenceStream | null;
  mapping_resolution_findings?: Soc2ResolutionFinding[];
};
